src/lua/log.lua

```diff
@@ -149,6 +149,8 @@ local function say(level, fmt, ...)
         for _, field in ipairs(special_fields) do
             fmt[field] = nil
         end
+        -- always encode tables as maps
+        setmetatable(fmt, json.map_mt)
```

Comment:
We set the metatable to the original object passed by a user. We shouldn't modify a user provided object.

We already add fields into the table, though, so it'll not get much worse. OTOH, this is explicitly defined as a problem, see #3853.

I would just add a `table.copy()`. It is simple and solves the problem. (And performance considerations are barely applicable if the behavior is incorrect.)

Reply:
Done

Reply:
AFAIU, now the assertion supposed to be solved by the following code:

```lua
-- set 'message' field if it is absent
if msg.message == nil then
    msg.message = msg[1]
    msg[1] = nil
end
```

However I guess it'll fail if `msg[1]` is `nil`:

```lua
tarantool> json.encode({nil, 2, 3})
---
- '[null,2,3]'
...
```

I would set the metatable unconditionally.

Reply:
NB: Verify the old/the new behavior on a table with __serialize or __tostring in its metatable. Whether table.copy() assigns the metatable to the copy? If so, shouldn't we assign our own one only if there is no original one? What if __serialize returns a string like '[1, 2, 3]' or even invalid json?

Reply:
I've addressed this in a new patch, could you please take a look? `__tostring` is not used during logging.

Reply:
Sorry, I missed your answer. Don't hesitate to ping me in a chat.

There are some doubts about current implementation, but I think that we should proceed now and maybe follow up something later.

For the record:

1. An object with `box.NULL` as a **key** (say `{[box.NULL] = 'x'}`) may pass the `next(msg) == nil` check. A correct check would use `type()`. However it doesn't matter much due to the next point.
2. There are objects that can't be encoded using just `json.encode(x)` (`userdata` for example). Those problems are usually solved using something like `local j = json.new() j.cfg({encode_use_tostring = true}) j.encode(x)`. However even this option doesn't help against `{[box.NULL] = 'x'}`.
3. The `__serialize` metamethod may return a table that is accessible for a user. We set a metatable to it and it may be considered as changing of a user provided object.
4. I'm not sure that removing of `msg[1]`, when it is placed to `msg.message`, is the good idea. What if a user pass an array intentionally? Maybe we should copy it into `msg.message`, not move. Maybe we should copy when the object is considered as an array, but move otherwise. Don't know.
5. I don't know, should we prefer `__tostring` or `__serialize`, when both are there. Now it is `__serialize` and it seems meaningful for a JSON logger, but I don't see a general rule.

I hope sooner or later we'll look into serialization and logging in a more general way. Given patch is tiny and solves several problems: I'm okay with it.

I would like someone else to look into the code and our discussion.